File: app/src/main/java/org/stypox/dicio/skills/telephone/ContactChooserIndex.java
package org.stypox.dicio.skills.telephone;

import org.dicio.numbers.util.Number;
import org.dicio.skill.Skill;
import org.dicio.skill.chain.InputRecognizer;

import java.util.List;

public class ContactChooserIndex extends Skill {
    private final List<NameNumberPair> contacts;
    private String input;
    private int index;

    ContactChooserIndex(final List<NameNumberPair> contacts) {
        this.contacts = contacts;
    }

    @Override
    public InputRecognizer.Specificity specificity() {
        return InputRecognizer.Specificity.high;
    }

    @Override
    public void setInput(final String theInput,
                         final List<String> inputWords,
                         final List<String> normalizedWordKeys) {
        this.input = theInput;
    }

    @Override
    public float score() {
        index = ctx().requireNumberParserFormatter()
                .extractNumbers(input)
                .preferOrdinal(true)
                .get().stream()
                .filter(Number.class::isInstance)
                .map(Number.class::cast)
                .filter(Number::isInteger)
                .mapToInt(number -> (int) number.integerValue())
                .findFirst()
                .orElse(0);

        return index <= 0 || index > contacts.size() ? 0.0f : 1.0f;
    }

    @Override
    public void processInput() {
    }

    @Override
    public void generateOutput() {
        if (index > 0 && index <= contacts.size()) {
            final NameNumberPair contact = contacts.get(index - 1);
            ConfirmCallOutput.callAfterConfirmation(this, contact.name, contact.number);
        }
    }

    @Override
    public void cleanup() {
        super.cleanup();
        input = null;
    }
}

File: app/src/main/java/org/stypox/dicio/skills/telephone/ContactChooserIndex.kt
package org.stypox.dicio.skills.telephone

import org.dicio.numbers.util.Number
import org.dicio.skill.Skill
import org.dicio.skill.chain.InputRecognizer.Specificity

class ContactChooserIndex internal constructor(private val contacts: List<NameNumberPair>) :
    Skill() {
    private var input: String? = null
    private var index = 0
    override fun specificity(): Specificity {
        return Specificity.high
    }

    override fun setInput(
        theInput: String,
        inputWords: List<String>,
        normalizedWordKeys: List<String>
    ) {
        input = theInput
    }

    override fun score(): Float {
        index = ctx().requireNumberParserFormatter()
            .extractNumbers(input)
            .preferOrdinal(true)
            .get().stream()
            .filter { obj: Any? -> Number::class.java.isInstance(obj) }
            .map { obj: Any? -> Number::class.java.cast(obj) }
            .filter { obj: Number -> obj.isInteger }
            .mapToInt { number: Number -> number.integerValue().toInt() }
            .findFirst()
            .orElse(0)
        return if (index <= 0 || index > contacts.size) 0.0f else 1.0f
    }

    override fun processInput() {}
    override fun generateOutput() {
        if (index > 0 && index <= contacts.size) {
            val contact = contacts[index - 1]
            ConfirmCallOutput.callAfterConfirmation(
                this,
                contact.name,
                contact.number
            )
        }
    }

    override fun cleanup() {
        super.cleanup()
        input = null
    }
}